File: world_simulation/sun.py
import time
class Sun:
    def __init__(self, shines = True, rise_time: int = 12, set_time: int = 12) -> None:
        self._rise_time = rise_time
        self._set_time = set_time
        self._set_start = 0
        self._rise_start = 0
        if shines:
            self.rise()
        else:
            self.set()
    
    def rise(self) -> None:
        self._rise_start = time.perf_counter()
        self._shines = True
    
    def set(self) -> None:
        self._set_start = time.perf_counter()
        self._shines = False
    
    def is_shining(self) -> bool:
        return self._shines
    
    def update_status(self) -> None:
        if self.is_shining():
            if time.perf_counter() - self._rise_start >= self._rise_time:
                self.set()
        else:
            if time.perf_counter() - self._set_start >= self._set_time:
                self.rise()
    
    def print_status(self):
        if self.is_shining():
            print(f"Sun is shining for {time.perf_counter() - self._rise_start:.0f} seconds.")
        else:
            print(f"Sun is set for {time.perf_counter() - self._set_start:.0f} seconds")

File: world_simulation_c/sun.c
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include "sun.h"

typedef struct Sun {
    int rise_time;
    int set_time;
    int rise_start;
    int set_start;
    int shines;
} Sun;

Sun * sun_init(){
    return (Sun*)malloc(sizeof(Sun));
}

void sun_construct(Sun * s){
    s->rise_time = 12;
    s->set_time = 12;
    rise(s);
}

void sun_destruct(Sun * s){
    free(s);
}

void rise(Sun * s){
    s->shines = 1;
    s->rise_start = time(NULL);
    
}

void set(Sun * s){
    s->shines = 0;
    s->set_start = time(NULL);
}

int is_shining(Sun * s){
    return s->shines;
}

void update_status(Sun * s){
    int current_time = time(NULL);
    if(s->shines){
        if(current_time - s->rise_start >= s->rise_time){
            set(s);
        }
    } else {
        if(current_time - s->set_start >= s->set_time){
            rise(s);
        }
    }
}

void sun_print_status(Sun * s){
    int current_time = time(NULL);
    if(s->shines){
        printf("Sun is shining for %d seconds.\n", current_time - s->rise_start);
    } else {
        printf("Sun is set for %d seconds.\n", current_time - s->set_start);
    }
}
